org.jacoco.tests/pom.xml

```diff
@@ -27,13 +27,12 @@
 
   <modules>
     <module>../org.jacoco.core.test</module>
+    <module>../org.jacoco.core.test.java5</module>
     <module>../org.jacoco.report.test</module>
     <module>../org.jacoco.agent.rt.test</module>
     <module>../org.jacoco.agent.test</module>
     <module>../org.jacoco.ant.test</module>
     <module>../org.jacoco.cli.test</module>
-    <module>../org.jacoco.examples.test</module>
```

Comment:
Why do we need to move these tests?

Reply:
Because otherwise they will be executed before modules from profiles.

Reply:
Sorry for my stupid question: And why would this be a problem?

Reply:
Because `org.jacoco.examples.test` and especially `jacoco-maven-plugin.test` are the slowest ones, while perform tests on a way higher level than `org.jacoco.core.test.*`.

Reply:
So you mean like "fail fast"? Where is that important? For local develop/test iterations I mostly run specific modules only anyways. For our CI all tests are executed anyways.

I just find it rather confusing that these modules are now included in specific JDK profiles.

Reply:
> So you mean like "fail fast"?

Yes.

> Where is that important?

For develop/test iterations I run specific tests in IDE, but then before commit I do verification build in command-line and it better fail on one of `org.jacoco.core.test.*` than on `jacoco-maven-plugin.test`. In later case, I'll be forced to manually select modules in command-line or in IDE.

> I just find it rather confusing that these modules are now included in specific JDK profiles.

I see only two options
1. add comment
2. introduce one more module with packaging `pom` to group `jdk` specific modules

Reply:
For clean dependencies I was thinking about option 2) anyways. Resulting in the following modules:

* `org.jacoco.core.test` - pure unit tests for `org.jacoco.core`
* `org.jacoco.core.test.validation` - base infrastructure used by the specific tests
* `org.jacoco.core.test.validation.xxx` - language version specific tests, depends on `org.jacoco.core.test.validation`

Does this make sense?

Reply:
Unfortunately we will anyway need one more module, because only modules with packaging `pom` can include other modules.

Reply:
Ok, I see. Then we would end up with something like:

* `org.jacoco.core.test.validation` - aggregator of type `pom`
* `org.jacoco.core.test.validation.base` - base infrastructure
* `org.jacoco.core.test.validation.xxx` - language version specific tests

Also looks good to me. Especially as we can separate the profile magic into the new module. Maybe it can also serve as a parent POM and we can e.g. move `jacoco.skip` to it.

Reply:
Re module names: For consistency we should also adjust Java package names to the new module names.

Reply:
I added `org.jacoco.core.test.validation` as parent for `org.jacoco.core.test.validation.*`

Not sure that I like that amount of folders in root. Maybe later we can think of grouping like for example

```
validation
  |- pom.xml
  |- java5
  |    \- pom.xml
  |- java7
  |    \- pom.xml
  \- java8
       \- pom.xml
```

Reply:
Absolutely! The same for the unit tests. We could simplify the structure if we could move to a more Maven standard project layout. But again... separate issue.

Reply:
> We could simplify the structure if we could move to a more Maven standard project layout.

In truly standard Maven layout, unit tests are located in the same module as main code, but we can't do this because we need to execute JaCoCo for JaCoCo itself - see https://github.com/jacoco/jacoco/issues/312#issuecomment-282526475

Reply:
I know 😉